File: backend/src/fetchers/fetch_parquet.ts
import fs from 'fs';
import os from 'os';
import path from 'path';
import * as https from 'node:https';
import { Config, Result } from '../index';
import { CustomOctokit } from '../lib/octokit';
import { queryRepoNames } from './fetcher_utils';

import { Database } from 'duckdb-async';

async function downloadParquetFile(url: string, outputPath: string) {
    return new Promise((resolve, reject) => {
        const file = fs.createWriteStream(outputPath);
        https.get(url, (response) => {
            if (response.statusCode !== 200) {
                reject(new Error(`Failed to get '${url}' (${response.statusCode})`));
                return;
            }
            response.pipe(file);
            file.on('finish', () => {
                file.close(resolve);
            });
        }).on('error', (err) => {
            fs.unlink(outputPath, () => reject(err));
        });
    });
}

export const addCondaData = async (result: Result, octokit: CustomOctokit, config: Config, startYear: number=2018) => {
    const repos = await queryRepoNames(octokit, config);
    const baseDir = path.join(os.homedir(), '.dashboard');

    const packages = repos.map((repo) => {return repo.name });

    const legacyPackagesMap = new Map<string, string>();
    if (config.organization === 'brainglobe') {
        const legacyPackages = JSON.parse(fs.readFileSync(path.resolve('brainglobe_legacy.json'), 'utf-8'));
        Object.entries(legacyPackages).forEach(([key, value]) => {
            if (typeof value === 'string') {
                legacyPackagesMap.set(key, value);
            }
        })

        packages.push(...legacyPackagesMap.keys());
    }

    if (!fs.existsSync(baseDir)) {
        fs.mkdirSync(baseDir);
    }

    let currYear = new Date().getFullYear();
    let lastMonth = 1;

    for (let i = startYear; i <= currYear; i++) {
        for (let j = 1; j <= 12; j++) {
            const fileName = path.join(baseDir, `${i}-${String(j).padStart(2, '0')}.parquet`);

            if (!fs.existsSync(fileName)) {
                const url = `https://anaconda-package-data.s3.amazonaws.com/conda/monthly/${i}/${i}-${String(j).padStart(2, '0')}.parquet`;

                // Check if the URL for the given month exists (status code 2xx)
                // The updates can take some time to be available, so there's
                // no easy way to know which month to stop at.
                const checkURLReq = await new Promise((resolve, reject) => {
                    fetch(url, {
                        method: "HEAD"
                    }).then(response => {
                        resolve(response.status.toString()[0] === "2")
                    }).catch(error => {
                        reject(false)
                    })
                })

                // If the URL does not exist, assume that there are no more
                // files to download and break the loop
                if (!checkURLReq) {
                    lastMonth = j > 1 ? j - 1 : 12;
                    // Make sure we break out of the outer loop as well
                    // Update teh current year to the last year we downloaded
                    currYear = i;
                    break;
                } else {
                    await downloadParquetFile(url, fileName);
                }
            }
        }
    }

    const db = await Database.create( `:memory:` );
    const formattedString = packages.map((pkg) => `'${pkg}'`).join(',');

    const totalDownloads = await db.all(`SELECT pkg_name, SUM(counts)::INTEGER AS total FROM '${baseDir}/*.parquet' WHERE pkg_name IN (${formattedString}) GROUP BY pkg_name`);

    if (lastMonth == 12) {
        currYear -= 1;
    }

    const lastMonthDownloads = await db.all(`SELECT pkg_name, SUM(counts)::INTEGER AS total FROM '${baseDir}/${currYear}-${String(lastMonth).padStart(2, '0')}.parquet' WHERE pkg_name IN (${formattedString}) GROUP BY pkg_name`);

    totalDownloads.forEach((row) => {
        if (legacyPackagesMap.has(row.pkg_name)) {
            row.pkg_name = legacyPackagesMap.get(row.pkg_name);
        }

        if ( !result.repositories[row.pkg_name].condaTotalDownloads ) {
            result.repositories[row.pkg_name].condaTotalDownloads = row.total;
        } else {
            result.repositories[row.pkg_name].condaTotalDownloads += row.total;
        }
    })

    lastMonthDownloads.forEach((row) => {
        if (legacyPackagesMap.has(row.pkg_name)) {
            row.pkg_name = legacyPackagesMap.get(row.pkg_name);
        }

        if (!result.repositories[row.pkg_name].condaMonthlyDownloads) {
            result.repositories[row.pkg_name].condaMonthlyDownloads = row.total;
        } else {
            result.repositories[row.pkg_name].condaMonthlyDownloads += row.total;
        }
    })

    return result;
};

File: lib/oss_dashboard/fetchers/fetch_parquet.py
"""Fetcher for Conda download statistics from AWS parquet files."""

import json
import logging
from datetime import datetime
from pathlib import Path

import duckdb
import requests

from oss_dashboard.constants import CHUNK_SIZE
from oss_dashboard.fetchers.utils import query_repo_names
from oss_dashboard.github_client import GitHubClient
from oss_dashboard.models import Config, Result

logger = logging.getLogger(__name__)


def _download_parquet_file(url: str, output_path: Path) -> bool:
    """Download a parquet file from URL.

    Args:
        url: URL to download from
        output_path: Path to save file

    Returns:
        True if download successful, False otherwise
    """
    try:
        response = requests.get(url, stream=True)
        if not response.ok:
            return False

        with open(output_path, "wb") as f:
            for chunk in response.iter_content(chunk_size=CHUNK_SIZE):
                f.write(chunk)
        return True
    except requests.RequestException as e:
        logger.warning("Error downloading %s: %s", url, e)
        return False


def _check_url_exists(url: str) -> bool:
    """Check if a URL exists.

    Args:
        url: URL to check

    Returns:
        True if URL exists (2xx status)
    """
    try:
        response = requests.head(url)
        return response.ok
    except requests.RequestException:
        return False


def add_conda_data(
    result: Result,
    client: GitHubClient,
    config: Config,
    start_year: int = 2018,
) -> Result:
    """Add Conda download statistics to the result.

    Args:
        result: Current result object
        client: GitHub client
        config: Configuration
        start_year: Year to start downloading data from

    Returns:
        Updated result with Conda download statistics
    """
    repos = query_repo_names(client, config)
    base_dir = Path.home() / ".dashboard"
    base_dir.mkdir(exist_ok=True)

    packages = [repo["name"] for repo in repos]

    # Handle legacy package mappings for brainglobe
    legacy_packages_map: dict[str, str] = {}
    if config.organization == "brainglobe":
        legacy_file = Path(__file__).parent / "brainglobe_legacy.json"
        if legacy_file.exists():
            with open(legacy_file) as f:
                legacy_packages = json.load(f)
                for key, value in legacy_packages.items():
                    if isinstance(value, str):
                        legacy_packages_map[key] = value
                packages.extend(legacy_packages_map.keys())

    curr_year = datetime.now().year
    last_month = 1

    # Download parquet files
    for year in range(start_year, curr_year + 1):
        for month in range(1, 13):
            file_name = base_dir / f"{year}-{month:02d}.parquet"

            if not file_name.exists():
                url = (f"https://anaconda-package-data.s3.amazonaws.com/"
                       f"conda/monthly/{year}/{year}-{month:02d}.parquet")

                # Check if URL exists
                if not _check_url_exists(url):
                    last_month = month - 1 if month > 1 else 12
                    curr_year = year
                    break

                logger.info("Downloading %s", url)
                if not _download_parquet_file(url, file_name):
                    logger.warning("Failed to download %s", url)
                    last_month = month - 1 if month > 1 else 12
                    curr_year = year
                    break
        else:
            continue
        break

    # Query with DuckDB
    con = duckdb.connect(":memory:")
    formatted_packages = ", ".join(f"'{pkg}'" for pkg in packages)

    try:
        # Get total downloads
        total_downloads = con.execute(
            f"""
            SELECT pkg_name, SUM(counts)::INTEGER AS total
            FROM '{base_dir}/*.parquet'
            WHERE pkg_name IN ({formatted_packages})
            GROUP BY pkg_name
            """
        ).fetchall()

        # Get last month downloads
        if last_month == 12:
            curr_year -= 1
            last_month = 12

        last_month_file = base_dir / f"{curr_year}-{last_month:02d}.parquet"
        if last_month_file.exists():
            last_month_downloads = con.execute(
                f"""
                SELECT pkg_name, SUM(counts)::INTEGER AS total
                FROM '{last_month_file}'
                WHERE pkg_name IN ({formatted_packages})
                GROUP BY pkg_name
                """
            ).fetchall()
        else:
            last_month_downloads = []

    except duckdb.Error as e:
        logger.error("Error querying parquet files: %s", e)
        return result
    finally:
        con.close()

    # Process total downloads
    for pkg_name, total in total_downloads:
        # Map legacy package names
        if pkg_name in legacy_packages_map:
            pkg_name = legacy_packages_map[pkg_name]

        if pkg_name not in result.repositories:
            continue

        result.repositories[pkg_name].conda_total_downloads += total

    # Process monthly downloads
    for pkg_name, total in last_month_downloads:
        # Map legacy package names
        if pkg_name in legacy_packages_map:
            pkg_name = legacy_packages_map[pkg_name]

        if pkg_name not in result.repositories:
            continue

        result.repositories[pkg_name].conda_monthly_downloads += total

    return result
